### DISCUSSION TITLE
Help : RHF usefield array data how to set data to different fields up on selection

### DISCUSSION THREAD

**Discussion by sunilbittu** [react-hook-form/react-hook-form#2621] [Q&A]
Hi, i have fieldarray with customer name (react-select dropdown), react-phone-input, input field.
upon selecting the option from customer name i would like to pre populate the email and phone number,
```
`const handleUserSelection = (val) => {
    console.log(val);
    const userObj = [];
    let userData = [];
    userData = customerList.filter(item => item.cid === val[0].value) ? customerData.filter(item => item.cid === val[0].value) : [];
    console.log(userData);
    if (userData && userData.length > 0) {
      setEmail(userData[0].cemail);
      setPhone(userData[0].cph);
   
    } else {
      setCName(val[0].value);
    }
  };
<label htmlFor="pkpName">Sender Name</label>
                            <Controller
                              control={control}
                              name={`pkp[${index}].name`}
                              defaultValue={cName}
                              render={props => (
                                <Select
                                  key={index}
                                  placeholder="Select Sender Name"
                                  onChange={handleUserSelection}
                                  selected={props.value}
                                  searchable
                                  create
                                  options={customerList}
                                />
                        )}
                            />`
```
Phone number is not setting for the first array its setting for the second array
any help with this.
thanks in advance.
https://codesandbox.io/s/react-hook-form-usefieldarray-111ly?file=/src/index.js

**Comment by bluebill1049** [comment]
```
onChange={(val)= > {
   props.onChange(val)
   handleUserSelection(val)
}}
```

try this.

**Comment by sunilbittu** [comment]
![Shyft-Club-Web_2](https://user-images.githubusercontent.com/10049424/90355699-ca18ce80-e06a-11ea-87f3-2b3d95de27a7.gif)
Happening like this

**Reply by bluebill1049** [reply]
oh sorry prepopulated! what about store those value and then use `append({ values })`

**Reply by sunilbittu** [reply]
what am trying to do is onChange getting the values and setting to a state and those values are showing in first field array (email and address) where as phone number is not shown.
but when i click on add another pickup (in directly calling append) it is showing phone number, email and address.
can i restrict setting values to one field array only

**Reply by bluebill1049** [reply]
when you click add another, are you using `append({})`?

**Reply by sunilbittu** [reply]
yes! something like this

```
const {
    fields: pickupFields, append: appendPickup, remove: removePickup,
  } = useFieldArray({
    control,
    name: 'pkp',
  });
onClick={() => { appendPickup(); }}
```

**Reply by bluebill1049** [reply]
```
onClick={() => { appendPickup({
  //data here? it should only fill up the newly added one
}); }}
```

**Reply by sunilbittu** [reply]
i think i have problem with react-phone-input one the first render its setting the value ` <Controller
                              as={PhoneInput}
                              control={control}
                              name={`pkp[${index}].ph`}
                              value={phone}
                              international
                              className="form-control"
                              rules={{ required: true }}
                              placeholder="Enter Mobile Number"
                            />`

**Reply by bluebill1049** [reply]
can you produce a simple codesandbox just for that input? i can take a look at it.

**Reply by sunilbittu** [reply]
its here
https://codesandbox.io/s/react-hook-form-usefieldarray-111ly?file=/src/index.js
but sandbox breaks when i add react-phone-input

**Reply by bluebill1049** [reply]
have you tried with `render`?

**Reply by sunilbittu** [reply]
`<Controller
                      control={control}
                      name={`pkp[${index}].ph`}
                      render={({ onChange, onBlur, value }) => {
                        return (
                          <PhoneInput
                            onChange={(e) => {
                              onChange(e.target.value);
                              // do you stuff here
                            }}
                            placeholder="Enter Mobile Number"
                            autoComplete="off"
                            international
                            value={phone}
                            className="form-control"
                          />
                        );
                      }}
                    />`

**Reply by sunilbittu** [reply]
strangely sandox breaks when i add react-phone-input

**Reply by sunilbittu** [reply]
made changes to react-phone-input version
now its displaying properly in the form.

**Reply by bluebill1049** [reply]
nice!

**Reply by sunilbittu** [reply]
@bluebill1049 every controller requires a defaultvalue right? but react-phone-input doesnt support the default value

getting a warning and value is not populated if i change value to defaultvalue.

![image](https://user-images.githubusercontent.com/10049424/90368527-ca739280-e087-11ea-949d-9bf276623193.png)

**Reply by bluebill1049** [reply]
not at the `render` level, you can remove the `defaultValue`

**Reply by sunilbittu** [reply]
yes my mistake.. its fine now.. on more thing
 `<h6 className="error-msg" style={{ color: 'red' }}> {errors && errors.pkp && errors.pkp.length > 0 && errors.pkp[index].email && 'Email Required'} </h6>`

is it proper way to check errors

**Reply by bluebill1049** [reply]
looks fine.

**Reply by bluebill1049** [reply]
by the way @sunilbittu you can use this as well: https://react-hook-form.com/api#ErrorMessage

**Reply by sunilbittu** [reply]
@bluebill1049 thanks refactored code with error messages api.

Is there any chance to set default value from a state to only one array in use field array

**Reply by bluebill1049** [reply]
```
reset({
   ...getValues(),
   // your data
})
```
When you got some free time, spend sometimes with the doc, just read through them all, trust me it will help.

**Reply by sunilbittu** [reply]
error messages are working fine but when i do but its minification is failing.

Failed to minify the code from this file:

        ./node_modules/react-hook-form/dist/index.esm.js:3

Read more here: http://bit.ly/2tRViJ9

import { ErrorMessage } from '@hookform/error-message';
import { Controller, useFieldArray, useForm } from 'react-hook-form/dist/index.ie11';

"@hookform/error-message": "0.0.4",
"react-hook-form": "^6.4.1",

**Reply by sunilbittu** [reply]
@bluebill1049 on form submit though I have value for phone number input it's showing error
https://codesandbox.io/s/react-hook-form-usefieldarray-forked-tqrt8

Select any name from sender name drop-down data will prepoulate after that onclick submit .. error is coming for phone input

**Reply by bluebill1049** [reply]
can you please follow up with steps to reproduce this problem?

**Reply by sunilbittu** [reply]
When ever we select a item from the sender name dropdown pre popultaing the fields like phone number and email.
where as phone number input is throwing required error.

**Reply by bluebill1049** [reply]
<img width="558" alt="Screen Shot 2020-08-24 at 9 14 45 pm" src="https://user-images.githubusercontent.com/10513364/91038859-d85c8100-e64e-11ea-9aef-ae7cba31e561.png">
i don't see errors?

**Reply by sunilbittu** [reply]
on submit its showing phone number is required

**Reply by bluebill1049** [reply]
<img width="1059" alt="Screen Shot 2020-08-24 at 9 58 51 pm" src="https://user-images.githubusercontent.com/10513364/91042390-01801000-e655-11ea-87d3-a0a44ee788d3.png">

**Reply by bluebill1049** [reply]
why you need that setValue there?

**Reply by sunilbittu** [reply]
added setvalue because i thought value from state is not setting to the phone input, even if i dont use set value, the error is same.

**Reply by bluebill1049** [reply]
no, you don't need setValue with Controller (Controller take care the registration and setValue), i removed it and it's not throwing error as my screenshot.

**Reply by sunilbittu** [reply]
ok sir, i'll try and update you in a min

**Reply by sunilbittu** [reply]
still its same sir
![image](https://user-images.githubusercontent.com/10049424/91043521-5ad64800-e631-11ea-8f80-157624e0b10b.png)

**Reply by sunilbittu** [reply]
![React-Hook-Form---useFieldArray-forked---CodeSandbox](https://user-images.githubusercontent.com/10049424/91061982-dcd26b00-e649-11ea-95ed-1d514805737d.gif)

**Reply by bluebill1049** [reply]
That's strange... i will take a look later today.

**Reply by bluebill1049** [reply]
<img width="1256" alt="Screen Shot 2020-08-25 at 10 22 26 am" src="https://user-images.githubusercontent.com/10513364/91108961-e0eba080-e6bc-11ea-9efe-a0ed5e12cf46.png">

now the phone error is going away, but whenever i select sender name it starts to wipe everything.

**Reply by bluebill1049** [reply]
you need to looking into that handle function : https://codesandbox.io/s/react-hook-form-usefieldarray-forked-7oxku?file=/src/index.js

**Reply by sunilbittu** [reply]
hi sir, handleuserselection method am setting values to the state and using them in controller.
`<Controller
                        control={control}
                        name={`pickupDtls[${index}].ph`}
                        defaultValue=""
                        render={({ onChange, onBlur, value }) => (
                          <PhoneInput
                            onChange={(val) => {
                              onChange(val);
                            }}
                            value={phone}
                            placeholder="Enter Mobile Number"
                            autoComplete="off"
                            international
                            className="form-control"
                          />
                        )}
                        rules={{ required: true }}
                      />` 
i tried doing with reset but not able to do.
const handleUserSelection = (val, index) => {
    console.log(index);
    const userObj = [];
    let userData = [];
    userData = customerList.filter((item) => item.value === val[0].value)
      ? customerList.filter((item) => item.value === val[0].value)
      : [];
    if (userData && userData.length > 0) {
      setCName(userData[0].label);
      setEmail(userData[0].cemail);
      setPhone(userData[0].cph);
      setLatitude(userData[0].jlat);
      setLongitude(userData[0].jlon);
      setPickupAddress(userData[0].jadd);
      userObj.push(userData[0]);
      setUserDetails(userDetails.concat(userObj));
      console.log(getValues().pickupDtls);
      // reset({
      //   ...getValues().pickupDtls,
      //   pickupDtls: [{ email: userData[0].cemail, ph: userData[0].cph }]
      // });
    } else {
      setCName(val[0].value);
    }
  };

**Reply by bluebill1049** [reply]
how come yesterday when i was playing with your CSB, this issue doesn't happen?

**Reply by sunilbittu** [reply]
am also wondering the same sir, something seriously wrong with my code.

**Reply by bluebill1049** [reply]
yea, perhaps revert back to what you had yesterday and go from there. i didn't fork your codesandbox, otherwise i could pass you the CSB.

**Reply by sunilbittu** [reply]
https://codesandbox.io/s/react-hook-form-usefieldarray-forked-tqrt8?file=/src/index.js:4507-4641

reverted changes

**Reply by bluebill1049** [reply]
looks like it's working now: https://codesandbox.io/s/react-hook-form-usefieldarray-forked-v37od

**Reply by bluebill1049** [reply]
you are missing the value prop

**Reply by sunilbittu** [reply]
i would like to set phone number depends on user selection. like email field, so am adding value={phone}
phone is coming from state

**Reply by bluebill1049** [reply]
that's the problem then, you shouldn't do that, Controller should all be wired up. you can use setValue to change the controller value tho.

**Reply by sunilbittu** [reply]
like in handleuserselection method i need to setValue(fieldname, phone) ?

**Reply by bluebill1049** [reply]
yes, is it not working for you?

**Reply by sunilbittu** [reply]
small confusion for me here, fieldnames writing here like  {`pickupDtls[${index}].ph`} how can i do that in setValue.
How to specify with array name?

**Reply by bluebill1049** [reply]
`setValue('pickupDtls[2].ph', 'value')`

**Reply by sunilbittu** [reply]
Focus seems to be undefined when we are using the controller, do we need to pass focus explicitly when using controller?

![image](https://user-images.githubusercontent.com/10049424/91256434-ed3b3080-e784-11ea-9988-510b818462e3.png)

**Reply by bluebill1049** [reply]
that's going to be the tricky part, I would recommend you do that after everything is working.

https://codesandbox.io/s/react-hook-form-controllerautofocus-v6-eeo66

**Reply by bluebill1049** [reply]
it's in the doc as well: https://react-hook-form.com/api#Controller

**Reply by sunilbittu** [reply]
following this video tutorial
https://www.youtube.com/watch?v=FNnSqP2rWis&feature=emb_logo
